File: 2048/main.py
import os
import random


def clear_console():
    os.system('cls' if os.name == 'nt' else 'clear')

def print_board(board):
    clear_console()
    for row in board:
        print('+----' * len(row) + '+')
        print('|'.join(f'{num:^5}' for num in row))
    print('+----' * len(board[0]) + '+')

def initialize_board():
    board = [[0] * 4 for _ in range(4)]
    add_random_tile(board)
    add_random_tile(board)
    return board

def add_random_tile(board):
    empty_cells = [(r, c) for r in range(4) for c in range(4) if board[r][c] == 0]
    if not empty_cells:
        return False
    row, col = random.choice(empty_cells)
    board[row][col] = 2 if random.random() < 0.9 else 4
    return True

def rotate_board(board):
    return [list(row) for row in zip(*board[::-1])]

def move_left(board):
    def merge(row):
        new_row = [i for i in row if i != 0]
        for i in range(len(new_row) - 1):
            if new_row[i] == new_row[i + 1]:
                new_row[i] *= 2
                new_row[i + 1] = 0
        new_row = [i for i in new_row if i != 0]
        new_row += [0] * (len(row) - len(new_row))
        return new_row

    return [merge(row) for row in board]

def move_right(board):
    return [row[::-1] for row in move_left([row[::-1] for row in board])]

def move_up(board):
    return rotate_board(move_left(rotate_board(rotate_board(rotate_board(board)))))

def move_down(board):
    return rotate_board(rotate_board(rotate_board(move_left(rotate_board(board)))))

def check_game_over(board):
    for row in board:
        if 0 in row:
            return False
    for row in board + rotate_board(board):
        for i in range(len(row) - 1):
            if row[i] == row[i + 1]:
                return False
    return True

def main():
    board = initialize_board()
    while True:
        print_board(board)
        move = input("Move (w/a/s/d): ").strip().lower()
        if move not in ['w', 'a', 's', 'd']:
            print("Invalid move. Use w, a, s, or d.")
            continue

        if move == 'w':
            new_board = move_up(board)
        elif move == 'a':
            new_board = move_left(board)
        elif move == 's':
            new_board = move_down(board)
        elif move == 'd':
            new_board = move_right(board)

        if new_board != board:
            board = new_board
            if not add_random_tile(board):
                print_board(board)
                print("Game Over!")
                break
        else:
            if check_game_over(board):
                print_board(board)
                print("Game Over!")
                break

if __name__ == "__main__":
    main()

File: 2048/main.js
const readline = require("readline");

// Clears the console in the CLI
function clearConsole() {
  console.clear();
}

// Print the current state of the board
function printBoard(board) {
  clearConsole();
  board.forEach((row) => {
    console.log("+----".repeat(row.length) + "+");
    console.log(
      "|" + row.map((num) => ` ${num === 0 ? " " : num}  `).join("|") + "|",
    );
  });
  console.log("+----".repeat(board[0].length) + "+");
}

// Initialize the board with two random tiles
function initializeBoard() {
  const board = Array.from({ length: 4 }, () => Array(4).fill(0));
  addRandomTile(board);
  addRandomTile(board);
  return board;
}

// Add a random tile (2 or 4) to an empty cell
function addRandomTile(board) {
  const emptyCells = [];
  for (let r = 0; r < 4; r++) {
    for (let c = 0; c < 4; c++) {
      if (board[r][c] === 0) emptyCells.push([r, c]);
    }
  }

  if (emptyCells.length === 0) return false;

  const [row, col] = emptyCells[Math.floor(Math.random() * emptyCells.length)];
  board[row][col] = Math.random() < 0.9 ? 2 : 4;
  return true;
}

// Rotate the board 90 degrees clockwise
function rotateBoard(board) {
  const newBoard = Array.from({ length: 4 }, () => Array(4).fill(0));
  for (let r = 0; r < 4; r++) {
    for (let c = 0; c < 4; c++) {
      newBoard[c][3 - r] = board[r][c];
    }
  }
  return newBoard;
}

// Handle left move logic
function moveLeft(board) {
  function merge(row) {
    const newRow = row.filter((val) => val !== 0);
    for (let i = 0; i < newRow.length - 1; i++) {
      if (newRow[i] === newRow[i + 1]) {
        newRow[i] *= 2;
        newRow[i + 1] = 0;
      }
    }
    const mergedRow = newRow.filter((val) => val !== 0);
    return mergedRow.concat(Array(4 - mergedRow.length).fill(0));
  }

  return board.map((row) => merge(row));
}

// Handle right move logic
function moveRight(board) {
  return board.map((row) => moveLeft([row.reverse()])[0].reverse());
}

// Handle up move logic
function moveUp(board) {
  const rotated = rotateBoard(board);
  return rotateBoard(rotateBoard(rotateBoard(moveLeft(rotated))));
}

// Handle down move logic
function moveDown(board) {
  const rotated = rotateBoard(board);
  return rotateBoard(rotateBoard(rotateBoard(moveRight(rotated))));
}

// Check if the game is over (no moves left)
function checkGameOver(board) {
  for (const row of board) {
    if (row.includes(0)) return false;
  }

  for (let r = 0; r < 4; r++) {
    for (let c = 0; c < 3; c++) {
      if (board[r][c] === board[r][c + 1] || board[c][r] === board[c + 1][r]) {
        return false;
      }
    }
  }

  return true;
}

// Main game loop
async function main() {
  let board = initializeBoard();

  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  while (true) {
    printBoard(board);

    const move = await new Promise((resolve) => {
      rl.question("Move (w/a/s/d): ", resolve);
    });

    if (!["w", "a", "s", "d"].includes(move)) {
      console.log("Invalid move. Use w, a, s, or d.");
      continue;
    }

    let newBoard;
    if (move === "w") newBoard = moveUp(board);
    else if (move === "a") newBoard = moveLeft(board);
    else if (move === "s") newBoard = moveDown(board);
    else if (move === "d") newBoard = moveRight(board);

    if (JSON.stringify(newBoard) !== JSON.stringify(board)) {
      board = newBoard;
      if (!addRandomTile(board)) {
        printBoard(board);
        console.log("Game Over!");
        break;
      }
    } else if (checkGameOver(board)) {
      printBoard(board);
      console.log("Game Over!");
      break;
    }
  }

  rl.close();
}

// Run the main game loop
main();
